src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/IQueueItem.cs

```diff
@@ -0,0 +1,34 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CommonLanguageServerProtocol.Framework;
+
+#nullable enable
+
+public interface IQueueItem<RequestContextType>
+{
+    /// <summary>
+    /// Begins executing the work specified by this queue item.
+    /// </summary>
+    Task CallbackAsync(RequestContextType? context, CancellationToken cancellationToken);
+
+    /// <inheritdoc cref="IRequestHandler.RequiresLSPSolution" />
+    bool RequiresLSPSolution { get; }
+
+    /// <inheritdoc cref="IRequestHandler.MutatesSolutionState" />
+    bool MutatesSolutionState { get; }
+
+    string MethodName { get; }
+
+    /// <summary>
+    /// The document identifier that will be used to find the solution and document for this request. This comes from the TextDocumentIdentifier returned from the handler itself via a call to <see cref="IRequestHandler{RequestType, ResponseType, TResponseContextType}.GetTextDocumentUri(RequestType)"/>.
+    /// </summary>
+    object? TextDocument { get; }
```

Comment:
Hmmm, could this be a `Uri?`?

Reply:
agreed.  we should make this as close to what lsp guarantees as possible.

Reply:
Just a small nit: the `Uri` class does not conform to the URI specification in some edge case scenarios.  (For example, `new Uri("http://*:5000", UriKind.Absolute)` will throw because the wildcard isn't supported even though this is RFC conformant.)  A `string` or other strongly typed document identifier might be a consideration.

Reply:
I thought this could just be a Uri too, but it turns out that Roslyn [relies on a vs-specific parameter](https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Protocol/Extensions/Extensions.cs#L76) that's added to `TextDocumentIdentifier` in order to identify files with the same name representing different solutions (I assume for net472 vs netstandard2.0 kind of stuff). If we just return a Uri here the test protecting that behavior fails, and we also lock ourselves away from any parameters which might be added to `TextDocumentIdentifier` in the future.

Reply:
Oh wow, good point! Is there a specific part of the project context that Roslyn requires / the API expects? Wondering if we can have a similar "GetProjectInformation" or something equivalent. /cc @dibarbet

Reply:
we need the VSTextDocumentIdentifier type which includes the project context.  We need all of the project context.
Project information is roslyn specific really though.

It might need to be a part of our Roslyn request handler API and context creation API.  the queue itself shouldn't need to know about it

Reply:
> It might need to be a part of our Roslyn request handler API and context creation API. the queue itself shouldn't need to know about it

Correct that the queue itself doesn't need to know about it, but the queue is what [initiates the creation of the RequestContext](https://github.com/dotnet/roslyn/pull/62924/files#diff-d247f5b12519875927ac9e8e65ce448ed8b611c5251d4dbeb068ee95ad03dffbR318), [which does need it](https://github.com/dotnet/roslyn/pull/62924/files#diff-05fa2dc8df33dfb4ecc0a6c1755d2be2b08b3293f61d29d0b867b1b5e60bad38R32)